src/App.tsx

```diff
@@ -15,36 +16,97 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+type SortType = 'length' | 'abc' | '';
+
+const getSortedGoods = (sortType: SortType, reversed: boolean): string[] => {
+  let sortedGoods = [...goodsFromServer];
+
+  if (sortType) {
+    sortedGoods.sort((a, b) => {
+      switch (sortType) {
+        case 'abc':
+          return a.localeCompare(b);
+        case 'length':
+          return a.length - b.length;
+        default:
+          return 0;
+      }
+    });
+  }
+
+  if (reversed) {
+    sortedGoods = sortedGoods.reverse();
+  }
+
+  return sortedGoods;
+};
+
 export const App: React.FC = () => {
+  const [sortType, setSortType] = useState<SortType>('');
+  const [reversed, setReversed] = useState<boolean>(false);
+
+  const visibleGoods = getSortedGoods(sortType, reversed);
+
+  const handleSort = (type: SortType): void => {
+    setSortType(type);
+  };
+
+  const handleReverse = (): void => {
+    setReversed(!reversed);
+  };
+
+  const handleReset = (): void => {
+    setReversed(false);
+    setSortType('');
+  };
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          type="button"
+          onClick={() => handleSort('abc')}
+          className={cn('button', 'is-info', {
+            'is-light': sortType !== 'abc',
+          })}
+        >
           Sort alphabetically
         </button>
 
-        <button type="button" className="button is-success is-light">
+        <button
+          type="button"
+          onClick={() => handleSort('length')}
+          className={cn('button', 'is-success', {
+            'is-light': sortType !== 'length',
+          })}
+        >
           Sort by length
         </button>
 
-        <button type="button" className="button is-warning is-light">
+        <button
+          type="button"
+          onClick={handleReverse}
+          className={cn('button', 'is-warning', { 'is-light': !reversed })}
+        >
           Reverse
         </button>
-
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {(sortType || reversed) && (
+          <button
+            type="button"
+            onClick={handleReset}
+            className="button is-danger is-light"
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {visibleGoods.map(good => (
+          <li data-cy="Good" key={good}>
```

Comment:
[CODE KNOWLEDGE] - When rendering a list, using the item itself as a `key` is not recommended if the items are not guaranteed to be unique. In this case, if `good` is not unique, consider using an index or a unique identifier as the `key` prop.